src/celeritas/ext/RootUniquePtr.hh

```diff
@@ -0,0 +1,70 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/RootUniquePtr.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+
+#include "celeritas_config.h"
+#include "corecel/Assert.hh"
+
+// Forward-declare ROOT
+class TObject;
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Helpers to prevent ROOT from propagating to downstream code.
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Call `TObject->Write()` before deletion. Used by TFile and TTree writer
+ * classes.
+ */
+template<class T>
+struct WriteRootDeleter
+{
+    void operator()(T* ptr);
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Deleter only. Used by TFile and TTree reader classes or TObjects that should
+ * not invoke `Write()`.
+ */
+template<class T>
+struct ReadRootDeleter
+{
+    void operator()(T* ptr);
+};
+
+//---------------------------------------------------------------------------//
+// Type aliases
+template<class T>
+using UPRootWriter = std::unique_ptr<T, WriteRootDeleter<T>>;
+
+template<class T>
+using UPRootReader = std::unique_ptr<T, ReadRootDeleter<T>>;
```

Comment:
```suggestion
using UPRootReader = std::unique_ptr<T>;
```
I am not quite sure why we even need the using, but maybe it helps clarifying intent?

Reply:
Yes, but it literally has nothing to do with ROOT.  i.e. you can/will use it from ANY of the library that we may depend on.  eg. Shouldn't we be using a similar part of some Geant4 things?  I.e the semantic is here is not "this is a ROOT object we need to do something different", it is "we want to reduce include parsing time, so we are adding an indirection for the destructor", so maybe something `IndirectDeleter`.

PS. So here we are exchanging developer time (faster compilation) for run-time (vaguely slower execution because of the indirection), it is likely the right choice here (very few of those deletion during execution) but just reminding myself of the trade-offs :)

Reply:
Well, this file instantiates the deleter only for ROOT types? The `WriteAndDeleter` definitely only applies to ROOT.

Also important to note in your PS: hiding the ROOT implementation in `.cc` files also allows us to declare ROOT a "private" dependency of Celeritas, so that we don't have to propagate its includes and library linkage downstream to parts of the code (e.g. unit tests) that don't use it.

Reply:
>  also allows us to declare ROOT a "private" dependency of Celeritas

Sure.  Is ROOT going to be the only such dependency?

Reply:
No, Geant4 has this as well (and maybe others will?). We have similar helper classes on an as-needed basis:
- GeantSetup.hh (`G4RunManager`)
- LoadGdml.hh (`G4VPhysicalVolume`)
- detail/VecgeomNavCollection.hh (`vecgeom::cxx::NavStatePool`)

Some of these have a little extra code in them (e.g., the run manager deleter prints a log message) but we could probably consolidate these and define with some macros? I like what you're getting at but I'm not sure of the tradeoff at this point for generalizing.

Reply:
I would just recommend to not over specify the names of the helper (i.e. `IndirectDeleter` is a `perfect` name even if you only use it for the ROOT unique pointer)

Reply:
Maybe `ExternDeleter` would be even better since `extern` is exactly the C++ semantics we want? @stognini could you change that for the cases that are templated and completely generic (i.e., that don't require any sort of `TObject` or call `Write`)?

Reply:
I like the `ExternDeleter` name. And should we worry about the filenames as well since we now have a generic deleter and a root-specific write-and-delete? Breaking into 2 files (one for the writer, and one for a generic safe unique ptr) seems too much, but it does make sense to me...

Reply:
> Maybe ExternDeleter would be even better since

I agree. :)

Reply:
> I like the `ExternDeleter` name. And should we worry about the filenames as well since we now have a generic deleter and a root-specific write-and-delete? Breaking into 2 files (one for the writer, and one for a generic safe unique ptr) seems too much, but it does make sense to me...

For now let's keep it together. In a separate PR we can consolidate the various "extern deleters".

Reply:
Okay. So I guess we can keep the `UPRootReadOnly` type alias for now until we consolidate different cases?

Reply:
That should probably also be `somethingExternDeleter`  ... 

Independently the current prefix always make me thing of the (related) project: https://uproot.readthedocs.io/en/latest/index.html

Reply:
We use `UP` uniformly as a prefix for `unique_ptr` (and `SP` for `shared_ptr`) so I think that's fine. Maybe `UPExtern`?

Reply:
I like `UPExtern`. It's concise and since it's templated it's always clear what we're using in any class.